File: src/main/java/me/sullivan/glasslang/parser/nodes/CallNode.java
package me.sullivan.glasslang.parser.nodes;

import me.sullivan.glasslang.interpreter.Interpreter;
import me.sullivan.glasslang.interpreter.primitives.Primitive;

import java.util.List;

public class CallNode extends Node
{

    private final Node atom;
    private final List<Node> argNodes;

    public CallNode(Node atom, List<Node> argNodes)
    {
        super(null);
        this.atom = atom;
        this.argNodes = argNodes;
    }

    @Override
    public String getValue()
    {
        return atom.getToken().getValue();
    }

    @Override
    public String toString()
    {
        return "CallNode(" + atom.getToken().getValue() + ", " + argNodes + ")";
    }

    @Override
    public Primitive<?> visitor(Interpreter interpreter)
    {
        return interpreter.visitCallNode(this);
    }

    public List<Node> getArgNodes()
    {
        return argNodes;
    }

    public Node nodeToCall()
    {
        return atom;
    }
}

File: src/me/sullivan/glasslang/parser/nodes/CallNode.kt
package me.sullivan.glasslang.parser.nodes

import me.sullivan.glasslang.interpreter.Interpreter
import me.sullivan.glasslang.interpreter.primitives.Primitive
import me.sullivan.glasslang.lexer.token.Token

class CallNode(token: Token, private val atom: Node, val argNodes: List<Node>) : Node(token)
{
    override val value: String
        get() = atom.token.value

    override fun toString(): String
    {
        return "CallNode($value, $argNodes)"
    }

    override fun visitor(interpreter: Interpreter): Primitive<*>
    {
        return interpreter.visitCallNode(this)
    }

    fun nodeToCall(): Node
    {
        return atom
    }
}
